### PR TITLE
Fixing potential crashes when loading  invalid KTX cached textures

### PR DESCRIPTION
**Pull request by samcake**
Adding a validation step at runtime for the cached KTX file in order to regenerate them if anything seems wrong.
This should prevent many of the crashes we are observing right now in the gl driver. This is happening when trying to load previously cached textures which are stored in the ktx_cache folder.
This pr add checks on the bytesize of each mips contained in the the file to validate that it s matching the expected size computed from the dimensions / format of the texture available in the Header.
If anything goes wrong, we simply bailout and regenerate the KTX texture (from the original source image) and overwrite the old cached ktx file with the new one.
We also fixed a problem with cube map in the way we were storing sizes in the KTX. The generated files are now valid compatible KTX that can be open with other viewers.

## TEST PLAN
To test this PR, you need to run the interface with different version of the KTX_cache folder in order to validate the new behavior.
The texture cache folder for a given version of High Fidelity on your computer is at the following appdata path:
 %UserName%/AppData/Local/%HighFidelityVersionName%/Interface/ktx_cache
For master or RC, HighFidelityVersionName is "High Fidelity"
For this PR, HighFidelityVersionName is "High Fidelity - PR10454"
HighFidelityVersionName is picked during the installation wizard, these are the name by default.

A ktx_cache folder with invalid ktx is available in this zip, it contains cached textures for dev-welcome.
1/ download this zip, open and rename the "ktx_cache_old" folder (containing all the ktx files) to "ktx_cache_invalid"
https://s3-us-west-1.amazonaws.com/hifi-content/chris/dev/office/ktx_cacheold.zip
2/ install this PR and current RC on your computer.
2 bis/ run both the PR and the RC and go to "dev-welcome" wait there  a few seconds for the content to be loaded and quit.
3/ Go in both appdata folders (for each versions) and rename the existing ktx_cache folder to "ktx_cache_ori"
4/ Copy in both appdata folders the downloaded ktx_cache_invalid (next to the existing one that you just renamed "ktx_cache_ori"
5/ Rename ktx_cache_invalid into ktx_cache (in both appdata folders)

At this point, you should be ready to run both master and this pr and spawn in "dev-welcome"
for each version, the ktx_cache contains the "invalid" cached ktx files.
In RC, it maybe run without any problem but will likely crash on some texture load/transfer. YOu should expect a crash.

In PR, this should run and not crash. Looking at the log, you should see a bunch of messages informing that 
"Invalid cached KTX url(...faultyTexture.png) under hash <HASHOFTHATFAULTYTEX> , recreating..."
ANd then a few lines later in the log you should see:
"Overwriting <HASHOFTHATFAULTYTEX>"
THis means that the invalid texture has been detected and has been recreated and stored in the ktx_cache folder replacing the faulty version.
After looking around the domain and loading most of the content you can quit.
Run the PR again and you shouldn't see any more "Invalid cached" message in the log at all (because it has been fixed in the previous run.

Once your done with all the testing, you can delete the ktx_cache folder in the appdata and replace it with the original one we renammed to ktx_cache_ori. simply rename it to ktx_cache.
Or just delete all these ktx_cache folders adn redownload everything fresh on the next run.

## TO DO
Need to update the Default skymap which is invalid right now...
THe default "blue horizon" skybox won't show because it's provided as a ktx file in the resources folder and it's invalid at this moment.

### CODE REVIEW

File: libraries/gpu/src/gpu/Texture.cpp [hunk 1]
```diff
@@ -441,7 +441,10 @@ void Texture::assignStoredMip(uint16 level, storage::StoragePointer& storage) {
     // THen check that the mem texture passed make sense with its format
     Size expectedSize = evalStoredMipSize(level, getStoredMipFormat());
     auto size = storage->size();
-    if (storage->size() <= expectedSize) {
+    if (storage->size() < expectedSize) {
```
Comment (jherico) [inline libraries/gpu/src/gpu/Texture.cpp]: the two blocks are identical.

File: libraries/gpu/src/gpu/Texture.cpp [hunk 2]
```diff
@@ -469,6 +472,8 @@ void Texture::assignStoredMipFace(uint16 level, uint8 face, storage::StoragePoin
     Size expectedSize = evalStoredMipFaceSize(level, getStoredMipFormat());
     auto size = storage->size();
     if (size <= expectedSize) {
+        _stamp++;
+    } else if (size == expectedSize) {
         _storage->assignMipFaceData(level, face, storage);
```
Comment (jherico) [inline libraries/gpu/src/gpu/Texture.cpp]: this block can't be reached.

File: libraries/gpu/src/gpu/Texture.cpp (hunk 1 above)
Comment (samcake) [inline libraries/gpu/src/gpu/Texture.cpp]: i separated the < and the == so i can breack point precisely when it s <

Comment (jherico) [inline libraries/gpu/src/gpu/Texture.cpp]: right, like I said, debug code.  if you want to leave it as is, you should add a comment that the two blocks are supposed to be identical.

File: libraries/ktx/src/ktx/KTX.cpp [hunk 3]
```diff
@@ -114,6 +114,9 @@ size_t Header::evalFaceSize(uint32_t level) const {
 }
 size_t Header::evalImageSize(uint32_t level) const {
     auto faceSize = evalFaceSize(level);
+    if ((faceSize < 4) || ((faceSize & 0x3) != 0)) {
```
Comment (jherico) [inline libraries/ktx/src/ktx/KTX.cpp]: you use this construct a number of times, but the second condition `faceSize & 0x3` automatically covers `faceSize < 4`, so the first condition is superfluous.

File: libraries/ktx/src/ktx/KTX.cpp [hunk 4]
```diff
@@ -139,6 +142,9 @@ ImageDescriptors Header::generateImageDescriptors() const {
     size_t imageOffset = 0;
     for (uint32_t level = 0; level < numberOfMipmapLevels; ++level) {
         auto imageSize = static_cast<uint32_t>(evalImageSize(level));
+        if ((imageSize < 4) || ((imageSize & 0x3) != 0)) {
+            return ImageDescriptors();
+        }
         if (imageSize == 0) {
             return ImageDescriptors();
```
Comment (jherico) [inline libraries/ktx/src/ktx/KTX.cpp]: This block is unreachable... the condition is covered by `imageSize < 4`

Comment (samcake) [inline libraries/ktx/src/ktx/KTX.cpp]: i think that works imageSize ending in 0x3 regardless of what s above the first 2 bits would trigger isn't it ?
Anyway i can rewrite yes

File: libraries/shared/src/shared/Storage.cpp [hunk 5]
```diff
@@ -21,6 +21,10 @@ ViewStorage::ViewStorage(const storage::StoragePointer& owner, size_t size, cons
 
 StoragePointer Storage::createView(size_t viewSize, size_t offset) const {
     auto selfSize = size();
+    if ((viewSize < 4) || ((viewSize & 0x3) != 0)) {
```
Comment (jherico) [inline libraries/shared/src/shared/Storage.cpp]: I don't think we should do this.  The storage abstraction isn't necessarily going to always be wrapping something that's 4 byte aligned.

Comment (samcake) [inline libraries/shared/src/shared/Storage.cpp]: well i think Storage should be actually aligned on 4bytes  but that s another debate
ok i ll remove

**Comment by hifi-gustavo** [comment]
Build console output for commit: 5bc8e098654bb0cbaa7897c02067278f38af4652 available at: https://gist.github.com/592be0767c0cfe3a86e3050392334554<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-5bc8e09.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-5bc8e09.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: bdb0414add26b3b792d9c1d16c36a933069acabb available at: https://gist.github.com/592be0767c0cfe3a86e3050392334554<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-bdb0414.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-bdb0414.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: d734358290cd4168591f1f4ac53d1471cb3b9e91 available at: https://gist.github.com/592be0767c0cfe3a86e3050392334554<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-d734358.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-d734358.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: c4e8885842ef82793227f1716887b20c41c28856 available at: https://gist.github.com/592be0767c0cfe3a86e3050392334554<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-c4e8885.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-c4e8885.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: f35b0297fa991ed2794500940aa53b69dbdda58f available at: https://gist.github.com/592be0767c0cfe3a86e3050392334554<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-f35b029.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-f35b029.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b38be7e561f5a2b30a8179cab40fc8626ac74b65 available at: https://gist.github.com/592be0767c0cfe3a86e3050392334554<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-b38be7e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-b38be7e.dmg

**Comment by YuppyPlays** [comment]
Hi there,

What's the test plan for this? ready to assist.

**Comment by samcake** [comment]
Hi @YuppyPlays See above, let me know if you have any questions

**Comment by hifi-gustavo** [comment]
Build console output for commit: b38be7e561f5a2b30a8179cab40fc8626ac74b65 available at: https://gist.github.com/592be0767c0cfe3a86e3050392334554<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-b38be7e.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR10454-b38be7e.dmg

**Comment by YuppyPlays** [comment]
Testing.

**Comment by YuppyPlays** [comment]
Testing complete:

Everything worked as expected

## Before KTX change:
PR & RC Both ran as expected
## After KTX change 1st run:
PR Worked but with the expected invalid ktx log messages, RC crashed as expected.
## After KTX change 2nd run (PR Only):
It was expected to see no invalid ktx messages though this one did pop up:

[05/13 22:01:37] [WARNING] [hifi.gpu-network] Invalid cached KTX  QUrl("http://hifi-content.s3.amazonaws.com/alan/dev/SKY-FacetLake_out.jpg")  under hash  4bbdf8f786470e4ab3e672d44b8e8df2 , recreating...

## After KTX change 3rd run (PR Only):
Following message appeared in the log.

[05/13 22:04:49] [WARNING] [hifi.gpu-network] Invalid cached KTX  QUrl("http://hifi-content.s3.amazonaws.com/alan/dev/SKY-FacetLake_out.jpg")  under hash  4bbdf8f786470e4ab3e672d44b8e8df2 , recreating...

Keep in mind for the 2nd and 3rd PR run i gave it more time to see if it would recreate, but it didn't seem to move after leaving it.


Not sure if pass?

**Comment by samcake** [comment]
Interesting, i tested myself and was not a problem ?
I think it s a pass in my opinion.
Happy to have one more qa

**Comment by CainFoool** [comment]
Testing

**Comment by CainFoool** [comment]
Test Complete

After ktx patching..

On PR10399: My game didn't crash but some textures simply didn't load, for example the table at dev-welcome. 

On PR10454: Similar experience, my game didn't crash but in this case the chairs around the table at dev-welcome didn't load but the table did this time. I did see logs for invalid KTX, recreating and upon re-launching the textures still weren't loading.

**Pull request merged**